File: dataAcq/buffer/csharp/WrappedObject.cs
using System;
using System.Collections;

namespace FieldTrip.Buffer
{
	public class WrappedObject
	{
        /// <summary>
        /// The type of data contained in the WrappedObject.
        /// </summary>
		public int Type{ get; set; }

        /// <summary>
        /// The number of elements.
        /// </summary>
		public int Numel{ get; set; }

        /// <summary>
        /// Size in bytes.
        /// </summary>
		public int Size{ get; set; }

        /// <summary>
        /// The data as an array.
        /// </summary>
		public object Array{ get; set; }

		public WrappedObject()
		{
			Type = DataType.UNKNOWN;
			Numel = 0;
			Size = 0;
			Array = null;
		}

		public WrappedObject(string s)
		{
			Type = DataType.CHAR;
			Numel = s.Length;
			Size = Numel;
			Array = s;
		}

		public WrappedObject(double x)
		{
			Type = DataType.FLOAT64;
			Numel = 1;
			Size = 8;
			Array = new double[] { x };
		}

		public WrappedObject(float x)
		{
			Type = DataType.FLOAT32;
			Numel = 1;
			Size = 4;
			Array = new float[] { x };
		}

		public WrappedObject(long x)
		{
			Type = DataType.INT64;
			Numel = 1;
			Size = 8;
			Array = new long[] { x };
		}

		public WrappedObject(int x)
		{
			Type = DataType.INT32;
			Numel = 1;
			Size = 4;
			Array = new int[] { x };
		}

		public WrappedObject(short x)
		{
			Type = DataType.INT16;
			Numel = 1;
			Size = 2;
			Array = new short[] { x };
		}

		public WrappedObject(byte x)
		{
			Type = DataType.INT8;
			Numel = 1;
			Size = 1;
			Array = new byte[] { x };
		}

        /// <summary>
        /// Create a wrapped object from a given object.
        /// </summary>
        /// <param name="obj"></param>
		public WrappedObject(object obj)
		{
	
			Type cls = obj.GetType();
			string name = cls.FullName;
	
			if (cls.IsArray) {
				Type elc = cls.GetElementType();
				if (!elc.IsPrimitive)
					return;
				
				if (name == "System.Double[]") {
					Type = DataType.FLOAT64;
					Array = ((double[])obj).Clone();
					Numel = ((double[])obj).Length;
				} else if (name == "System.Single[]") {
					Type = DataType.FLOAT32;
					Array = ((float[])obj).Clone();
					Numel = ((float[])obj).Length;
				} else if (name == "System.Int64[]") {
					Type = DataType.INT64;
					Array = ((long[])obj).Clone();
					Numel = ((long[])obj).Length;
				} else if (name == "System.Int32[]") {
					Type = DataType.INT32;
					Array = ((int[])obj).Clone();
					Numel = ((int[])obj).Length;
				} else if (name == "System.Int16[]") {
					Type = DataType.INT16;
					Array = ((short[])obj).Clone();
					Numel = ((short[])obj).Length;
				} else if (name == "System.Byte[]") {
					Type = DataType.INT8;
					Array = ((byte[])obj).Clone();
					Numel = ((byte[])obj).Length;
				} else {
					return; // keep as unknown
				}
				Size = Numel * DataType.wordSize[Type];
				return;
			} else if (name == "System.String") {
				Type = DataType.CHAR;
				Array = obj;
				Numel = ((string)obj).Length;
				Size = Numel;
				return;
			} else if (name == "System.Double") {
				Type = DataType.FLOAT64;
				Array = new double[] { ((double)obj) };
			} else if (name == "System.Single") {
				Type = DataType.FLOAT32;
				Array = new float[] { ((float)obj) };
			} else if (name == "System.Int64") {
				Type = DataType.INT64;
				Array = new long[] { ((long)obj) };
			} else if (name == "System.Int32") {
				Type = DataType.INT32;
				Array = new int[] { ((Int32)obj) };
			} else if (name == "System.Int16") {
				Type = DataType.INT16;
				Array = new short[] { ((short)obj) };
			} else if (name == "System.Byte") {
				Type = DataType.INT8;
				Array = new byte[] { ((byte)obj) };		
			} else {
				return;
			}
			Numel = 1;
			Size = DataType.wordSize[Type];
		}

        /// <summary>
        /// Serialize the WrappedObject to the specified <see cref="FieldTrip.Buffer.ByteBuffer"/>.
        /// </summary>
        /// <param name="buf">The buffer to serialize to.</param>
		public void Serialize(ByteBuffer buf)
		{
			switch (Type) {
				case DataType.CHAR:
					buf.PutString(Array.ToString());
					break;
				case DataType.UINT8:
				case DataType.INT8:
					buf.Put((byte[])Array);
					break;
				case DataType.UINT16:
				case DataType.INT16:
					buf.AsShortBuffer().Put((short[])Array);
					break;
				case DataType.UINT32:
				case DataType.INT32:
					buf.AsIntBuffer().Put((int[])Array);
					break;
				case DataType.UINT64:
				case DataType.INT64:
					buf.AsLongBuffer().Put((long[])Array);
					break;
				case DataType.FLOAT32:
					buf.AsFloatBuffer().Put((float[])Array);
					break;
				case DataType.FLOAT64:
					buf.AsDoubleBuffer().Put((double[])Array);
					break;
			}
		}

        /// <summary>
        /// Creates a string representation of the data.
        /// </summary>
        /// <returns>The string representation of the data.</returns>
		public override string ToString()
		{
			if (Type == DataType.CHAR)
				return (string)Array;
			if (Type == DataType.FLOAT64)
				return (((double[])Array)[0]).ToString();
			if (Type == DataType.FLOAT32)
				return (((float[])Array)[0]).ToString();
			if (Type == DataType.INT64)
				return (((long[])Array)[0]).ToString();
			if (Type == DataType.INT32)
				return (((int[])Array)[0]).ToString();
			if (Type == DataType.INT16)
				return (((short[])Array)[0]).ToString();
			if (Type == DataType.INT8)
				return (((byte[])Array)[0]).ToString();
			return Array.ToString();
		}
	}
}

File: java/SignalViewer/AndroidStudioProjects/SignalViewer/app/src/main/java/nl/fcdonders/fieldtrip/bufferclient/WrappedObject.java
/*
 * Copyright (C) 2010, Stefan Klanke
 * Donders Institute for Donders Institute for Brain, Cognition and Behaviour,
 * Centre for Cognitive Neuroimaging, Radboud University Nijmegen,
 * Kapittelweg 29, 6525 EN Nijmegen, The Netherlands
 */
package nl.fcdonders.fieldtrip.bufferclient;

import java.nio.*;

/** A class for wrapping relevant Java objects in a way that
	is easily convertible to FieldTrip data types.
*/
public class WrappedObject {
	protected int type;
	protected int numel;
	protected int size;
	protected Object array;
	
	public WrappedObject() {
		type  = DataType.UNKNOWN;
		numel = 0;
		size  = 0;
		array = null;
	}
	
	public WrappedObject(String s) {
		type  = DataType.CHAR;
		numel = s.getBytes().length;
		size  = numel;
		array = s;
	}
	
	public WrappedObject(double x) {
		type  = DataType.FLOAT64;
		numel = 1;
		size  = 8;
		array = new double[] {x};
	}	
	
	public WrappedObject(float x) {
		type  = DataType.FLOAT32;
		numel = 1;
		size  = 4;
		array = new float[] {x};
	}
	
	public WrappedObject(long x) {
		type  = DataType.INT64;
		numel = 1;
		size  = 8;
		array = new long[] {x};
	}	
	
	public WrappedObject(int x) {
		type  = DataType.INT32;
		numel = 1;
		size  = 4;
		array = new int[] {x};
	}
	
	public WrappedObject(short x) {
		type  = DataType.INT16;
		numel = 1;
		size  = 2;
		array = new short[] {x};
	}
	
	public WrappedObject(byte x) {
		type  = DataType.INT8;
		numel = 1;
		size  = 1;
		array = new byte[] {x};
	}	

	public WrappedObject(boolean x) {
		type  = DataType.INT8;
		numel = 1;
		size  = 1;
		array = new byte[]{(byte)(x?1:0)};
	}
	
	public WrappedObject(Object obj) {
		this();
	
		Class cls = obj.getClass();
		String name = cls.getName();
		//System.out.println("cls="+cls+" name=" + name);
		
		if (cls.isArray()) {
			Class elc = cls.getComponentType();
			
         if ( name == null ) {
             return; // keep as unknown

         }else if   (name.equals("[D")) {
				type = DataType.FLOAT64;
				array = ((double[]) obj).clone();
				numel = ((double[]) obj).length;

			} else if  (name.equals("[Ljava.lang.Double") || name.equals("[Ljava.lang.Double;") ) {
				type = DataType.FLOAT64;
				double[] tmp = new double[((Double[])obj).length];
				for ( int i=0; i<tmp.length; i++ ) tmp[i] = (double) ((Double[])obj)[i];
				array = tmp;
				numel = tmp.length;

			} else if (name.equals("[F")) {
				type = DataType.FLOAT32;
				array = ((float[]) obj).clone();
				numel = ((float[]) obj).length;

			} else if  (name.equals("[Ljava.lang.Float") || name.equals("[Ljava.lang.Float;")) {
				type = DataType.FLOAT32;
				float[] tmp = new float[((Float[])obj).length];
				for ( int i=0; i<tmp.length; i++ ) tmp[i] = (float) ((Float[])obj)[i];
				array = tmp;
				numel = tmp.length;

			} else if (name.equals("[J")) {
				type = DataType.INT64;
				array = ((long[]) obj).clone();
				numel = ((long[]) obj).length;

			} else if (name.equals("[I")) {
				type = DataType.INT32;
				array = ((int[]) obj).clone();
				numel = ((int[]) obj).length;
			
			} else if  (name.equals("[Ljava.lang.Integer") || name.equals("[Ljava.lang.Integer;")) {
				type = DataType.INT32;
				int[] tmp = new int[((Integer[])obj).length];
				for ( int i=0; i<tmp.length; i++ ) tmp[i] = (int) ((Integer[])obj)[i];
				array = tmp;
				numel = tmp.length;

			} else if (name.equals("[S")) {
				type = DataType.INT16;
				array = ((short[]) obj).clone();
				numel = ((short[]) obj).length;
			
			} else if (name.equals("[B")) {
				type = DataType.INT8;
				array = ((byte[]) obj).clone();
				numel = ((byte[]) obj).length;
			
			} else if  (name.equals("[Ljava.lang.Byte") || name.equals("[Ljava.lang.Byte;") ) {
				type = DataType.INT8;
				double[] tmp = new double[((Boolean[])obj).length];
				for ( int i=0; i<tmp.length; i++ ) tmp[i] = (byte)((Byte[])obj)[i];
				array = tmp;
				numel = tmp.length;

			} else if (name.equals("[Z")) { // boolean
				type = DataType.INT8;
				byte[] tmp = new byte[((boolean[])obj).length];
				for ( int i=0; i<tmp.length; i++ ) tmp[i] = (byte)((((boolean[])obj)[i])?1:0);
				array = tmp;
				numel = tmp.length;

			} else if  (name.equals("[Ljava.lang.Boolean") || name.equals("[Ljava.lang.Boolean;") ) {
				type = DataType.INT8;
				double[] tmp = new double[((Boolean[])obj).length];
				for ( int i=0; i<tmp.length; i++ ) tmp[i] = (byte)((((Boolean[])obj)[i])?1:0);
				array = tmp;
				numel = tmp.length;

			} else {
				return; // keep as unknown
			}
			size  = numel * DataType.wordSize[type];
			return;
		} else if (name.equals("java.lang.String")) {
			type = DataType.CHAR;
			array = obj;
			numel = ((String) obj).getBytes().length;
			size  = numel;
			return;
		} else if (name.equals("java.lang.Double")) {
			type = DataType.FLOAT64;
			array = new double[] {((Double) obj).doubleValue()};
		} else if (name.equals("java.lang.Float")) {
			type = DataType.FLOAT32;
			array = new float[] {((Float) obj).floatValue()};
		} else if (name.equals("java.lang.Long")) {
			type = DataType.INT64;
			array = new long[] {((Long) obj).longValue()};
		} else if (name.equals("java.lang.Integer")) {
			type = DataType.INT32;
			array = new int[] {((Integer) obj).intValue()};
		} else if (name.equals("java.lang.Short")) {
			type = DataType.INT16;
			array = new short[] {((Short) obj).shortValue()};
		} else if (name.equals("java.lang.Byte")) {
			type = DataType.INT8;
			array = new byte[] {((Byte) obj).byteValue()};		
		} else if (name.equals("java.lang.Boolean")) {
			type = DataType.INT8;
			array = new byte[] {(byte)(((Boolean) obj).booleanValue()?1:0)};		
		} else {
			return;
		}
		numel = 1;
		size  = DataType.wordSize[type];
	}	
		
	public void serialize(ByteBuffer buf) {
		switch(type) {
			case DataType.CHAR:
				buf.put(((String) array).getBytes());
				break;
			case DataType.UINT8:
			case DataType.INT8:
				buf.put((byte[]) array);
				break;
			case DataType.UINT16:
			case DataType.INT16:
				buf.asShortBuffer().put((short[]) array);
				break;
			case DataType.UINT32:
			case DataType.INT32:
				buf.asIntBuffer().put((int[]) array);
				break;
			case DataType.UINT64:
			case DataType.INT64:
				buf.asLongBuffer().put((long[]) array);
				break;
			case DataType.FLOAT32:
				buf.asFloatBuffer().put((float[]) array);
				break;
			case DataType.FLOAT64:
				buf.asDoubleBuffer().put((double[]) array);
				break;
		}
	}	
	
	 // methods to extract info from the data
	 public int getType() { return type; }
	 public Object getArray() { return array; }

	public String toString() {
		 String str=new String();
		if (type == DataType.CHAR) return (String) array;
		if (type == DataType.FLOAT64) {
			 str = String.valueOf(((double[]) array)[0]);
			 for ( int i=1; i<numel; i++ ){
				  str = str + "," + String.valueOf(((double[]) array)[i]);
			 }
			 return str;
		}
		if (type == DataType.FLOAT32) {
			 str = String.valueOf(((float[]) array)[0]);
			 for ( int i=1; i<numel; i++ ){
				  str = str + "," + String.valueOf(((float[]) array)[i]);
			 }
			 return str;
		}
		if (type == DataType.INT64) {
			 str = String.valueOf(((long[]) array)[0]);
			 for ( int i=1; i<numel; i++ ){
				  str = str + "," + String.valueOf(((long[]) array)[i]);
			 }
			 return str;
		}
		if (type == DataType.INT32) {
			 str = String.valueOf(((int[]) array)[0]);
			 for ( int i=1; i<numel; i++ ){
				  str = str + "," + String.valueOf(((int[]) array)[i]);
			 }
			 return str;
		}
		if (type == DataType.INT16) {
			 str = String.valueOf(((short[]) array)[0]);
			 for ( int i=1; i<numel; i++ ){
				  str = str + "," + String.valueOf(((short[]) array)[i]);
			 }
			 return str;
		}
		if (type == DataType.INT8) {
			 str = String.valueOf(((byte[]) array)[0]);
			 for ( int i=1; i<numel; i++ ){
				  str = str + "," + String.valueOf(((byte[]) array)[i]);
			 }
			 return str;
		}
		return array.toString();
	}
}
